AX: CSS content property replacement text ignored when it is the empty string

annevk opened the pull request:
#### 93c3eb8ac99a6b6e3894deaf14e5d38304898b48
<pre>
AX: CSS content property replacement text ignored when it is the empty string
<a href="https://bugs.webkit.org/show_bug.cgi?id=270377">https://bugs.webkit.org/show_bug.cgi?id=270377</a>
<a href="https://rdar.apple.com/123919677">rdar://123919677</a>

Reviewed by Antoine Quint.

This issue came to light as part of the &lt;input type=checkbox switch&gt;
demo, but is more universally applicable as shown in the tests.

While here, also cleanup some of the code as the render objects
involved cannot become null.

New web-platform-tests tests are exported:
<a href="https://github.com/web-platform-tests/wpt/pull/44971">https://github.com/web-platform-tests/wpt/pull/44971</a>

* LayoutTests/accessibility/mac/alt-for-css-content-expected.txt:
* LayoutTests/accessibility/mac/alt-for-css-content.html:
* LayoutTests/imported/w3c/web-platform-tests/accname/name/comp_name_from_content-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/accname/name/comp_name_from_content.html:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textUnderElement const):

Canonical link: <a href="https://commits.webkit.org/275872@main">https://commits.webkit.org/275872@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2f5056817c8fd61f80d5cf9f7bd04c5cbff785a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43011 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22037 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45413 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45635 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39136 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45317 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25762 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19460 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35565 "Found 1 new test failure: http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43584 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19070 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37040 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16560 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16662 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38098 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1067 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39193 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47167 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17858 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14714 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42343 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19463 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37363 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40996 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9595 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19642 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19093 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -718,7 +718,7 @@ String AccessibilityRenderObject::textUnderElement(AccessibilityTextUnderElement
                 // The alt attribute may be set on a text fragment through CSS, which should be honored.
                 const auto& altText = renderTextFragment->altText();
-                if (!altText.isEmpty())
+                if (!altText.isNull())
```
This touches only `RenderTextFragment`, does the case with `content: url(...) / ""` work as expected?

I'm seeing other isEmpty() checks: https://searchfox.org/wubkat/search?q=AltText.isEmpty&path=&case=false&regexp=false

lukewarlow replied:
Most of those results seem to be related to the img alt attribute rather than CSS content alt text fwiw

nt1m replied:
I do want to make sure `content: url(...) / ""` also works properly, because I'm fairly sure that doesn't use `RenderTextFragment`

cookiecrook replied:
Are there any changes to the result of [this WPT accname "name form contents" test file](https://wpt.fyi/results/accname/name/comp_name_from_content.html?label=master&label=experimental&aligned&q=label%3Aaccessibility)? If not, will you add a new [interop-accessibility issue](https://github.com/web-platform-tests/interop-accessibility/issues) that we probably need to add a few more subtests? Or you could add some as part of this PR. It seems like nullability of the pseudo element label was not otherwise tested in WPT.

nt1m reviewed (commented):
I would augment LayoutTests/imported/w3c/web-platform-tests/accname/name/comp_name_from_content.html

graouts commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1842,10 +1842,10 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
     if (usesAltTagForTextComputation()) {
         if (auto* renderImage = dynamicDowncast<RenderImage>(renderer())) {
-            String renderAltText = renderImage->altText();
+            auto renderAltText = renderImage->altText();
```
I prefer `auto&` in this case since we're dealing with a reference.

graouts commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -717,13 +717,11 @@ String AccessibilityRenderObject::textUnderElement(AccessibilityTextUnderElement
                 // The alt attribute may be set on a text fragment through CSS, which should be honored.
-                const auto& altText = renderTextFragment->altText();
-                if (!altText.isEmpty())
+                if (const auto& altText = renderTextFragment->altText(); !altText.isNull())
```
I don't think the `const` is useful here, `auto&` would suffice.

graouts commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -717,13 +717,11 @@ String AccessibilityRenderObject::textUnderElement(AccessibilityTextUnderElement
                 // The alt attribute may be set on a text fragment through CSS, which should be honored.
-                const auto& altText = renderTextFragment->altText();
-                if (!altText.isEmpty())
+                if (auto& altText = renderTextFragment->altText(); !altText.isNull())
```
I'm not 100% sure, but given `String` has its negation operator defined as follows:

```
inline bool operator!(const String& string) { return string.isNull(); }
```

Maybe you don't need the second clause in that if statement?

annevk replied:
I suspect you're correct, but I thought that for strings we wanted these checks to be explicit as it can be little unclear otherwise whether you meant to include empty strings or not?

annevk replied:
I'm going to leave this explicit for now. But happy to fix as a follow-up.

annevk committed:
AX: CSS content property replacement text ignored when it is the empty string
https://bugs.webkit.org/show_bug.cgi?id=270377
rdar://123919677

Reviewed by Antoine Quint.

This issue came to light as part of the <input type=checkbox switch>
demo, but is more universally applicable as shown in the tests.

While here, also cleanup some of the code as the render objects
involved cannot become null.

New web-platform-tests tests are exported:
https://github.com/web-platform-tests/wpt/pull/44971

* LayoutTests/accessibility/mac/alt-for-css-content-expected.txt:
* LayoutTests/accessibility/mac/alt-for-css-content.html:
* LayoutTests/imported/w3c/web-platform-tests/accname/name/comp_name_from_content-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/accname/name/comp_name_from_content.html:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textUnderElement const):

Canonical link: https://commits.webkit.org/275872@main
